### DISCUSSION TITLE
UX - Context menus when right-clicking

### DISCUSSION THREAD

**Discussion by jounih** [audacity/audacity#867] [Ideas & Feature Requests]
Right now, only created labels have context menus on right click.

I propose to add this behaviour to audio clips and empty regions as well (and possibly other areas where it makes sense).

This would be an addition, not removing any existing behaviour - the top menus and keyboard shortcuts would still also work just as they do now.

The benefits of context menus are many:

- Easier discovery of functionality, in the context you want to perform an action (ie. right click where you want to split a track, create a new track by right clicking on an empty area and selecting "Add track..." etc)
- Fast access to relevant actions without having to memorise keyboard shortcuts (although those still work fine)
- Expected behaviour from most other applications

Below are a couple of mockups (the exact content of the context menus should be decided after polling users what the most frequent/most useful actions would be).

<img width="1920" alt="1 0 audacity3 menu on empty@2x" src="https://user-images.githubusercontent.com/10109867/117790900-b25d2800-b241-11eb-96e2-3ddd79e1f555.png">

<img width="1920" alt="1 0 audacity3 menu on audio clip@2x" src="https://user-images.githubusercontent.com/10109867/117790909-b4bf8200-b241-11eb-840d-485019994add.png">

In terms of technical implementation, Audacity already has (native) context menus on selected label clips - these could be built the same way.

Thoughts?

**Comment by SteveDaulton** [comment]
The question is: what behaviours / actions / commands to include in a right click menu.
Intuitively I'd expect to see the basic edit commands: Cut, Delete, Copy, Paste.
The danger is that the right click menu becomes overloaded with so many options that it becomes no more useful than simply using the main menus (with the possible exception of macOS due to it's dreadful main menu layout detached from the app).

**Reply by jounih** [reply]
@SteveDaulton Fully agree - the context menus should be kept lightweight, with the most common actions - and not overloaded. This doesn't mean however, that we shouldn't add them at all - as long as we can keep them lightweight, the benefits are plenty!

I don't think there are any statistics in terms of what features users use the most - I can set up a public poll so we get some structured data on what actions users would like to see the most?

**Reply by SteveDaulton** [reply]
The problem with straw polls is that the answer is overwhelmingly biased by the exact wording of the question, and who you ask. It's virtually impossible to achieve a representative poll without a lot of resources and  careful statistical analysis. If you ask here on GitHub you are likely to get completely different answers than if you ask on the Audacity help forum, and completely different answers if you ask on a facebook page.
This is an example of where telemetry data, which has been emphatically rejected on GitHub for Audacity, becomes invaluable.

**Reply by jounih** [reply]
@SteveDaulton agreed, however in the absence of analytics, polling is as good as it gets in terms of structured data- and the poll wording can be kept neutral so as not to introduce bias. The idea is to list many common contextual actions in the poll, and see which ones get the highest percentage. We should publish a link to the poll in many places - the forum, github, Audacity blog and Facebook to get a wide distribution of replies.

**Reply by petersampsonaudacity** [reply]
We have had user requests in that past for more right-click context menus,  Se this section of Wiki>Fearure Requests:  https://wiki.audacityteam.org/wiki/Feature_Requests#contextmenu

**Reply by petersampsonaudacity** [reply]
See also this proposal in the Wiki:  https://wiki.audacityteam.org/wiki/Proposal_Smart_Help

**Reply by jounih** [reply]
> We have had user requests in that past for more right-click context menus, Se this section of Wiki>Fearure Requests: https://wiki.audacityteam.org/wiki/Feature_Requests#contextmenu

Thanks Peter this is very useful! These are a good set of options for a poll. How did you collect votes for these in the past?

**Reply by petersampsonaudacity** [reply]
Those votes were mostly collected from the Forum (with some from the now defunct feedback@ email channel)

**Reply by stmueller** [reply]
I'll point out that audacity has no/few context menus because of the lack of a second mouse buttons on macs. Yes this is still accessible with some keyboard-click or if people buy a real mouse, but it means you cannot count on the function being easily accessed for a large number of users (including windows users who never think to right-click), and so it must be supported in another way as well.  Oftentimes this meant simplifying things, but also prevented cramming the entire menu into a context menu which might be daunting, and we've all seen context menus in open source products that make things worse, not better.  That doesn't mean a context menu could not improve workflow, especially for advanced users.  According to various tutorial videos I have seen, power users mostly just use keyboard shortcuts for the things that would go into this menu.  A reasonable policy might be that anything in the context menu needs to be accessible elsewhere, and things should only be in a context menu if they specifically apply to the current selection or playback position (e.g., split track, cut,copy, duplicate paste, silence, effects).

**Reply by jounih** [reply]
All the native MacOS software I use has context menus - you can set it up on the Mac trackpad with two fingers (same on Magic mouse if you don't have a trackpad).

https://support.apple.com/en-gb/guide/mac-help/mh35853/mac

I think the trick is, as Steve said it, to make sure it's not crammed too full and has only the often used actions.

I'll reiterate that this should be an additional way to access often used actions - not a replacement. The actions would still work with the same keyboard shortcuts, and main menu items as today. Context menus just provide an additional, easily discoverable way to access quick actions, in context.
